File: src/RentCars.Tools/Results/DataResult.cs
using System.Diagnostics.CodeAnalysis;

namespace RentCars.Tools.Results;

public class DataResult<T>
{
    public T? Data { get; }
    public String[] Errors { get; }

    [MemberNotNullWhen(true, nameof(Data))]
    public Boolean IsSuccess => Errors.Length == 0;

    private DataResult(T? data, String[] errors)
    {
        Data = data;
        Errors = errors;
    }

    public static DataResult<T> Success(T data)
    {
        return new DataResult<T>(data, Array.Empty<String>());
    }

    public static DataResult<T> Fail(String error)
    {
        return new DataResult<T>(default, new[] { error });
    }

    public static DataResult<T> Fail(String[] errors)
    {
        return new DataResult<T>(default, errors);
    }
}

File: API/api.client/src/tools/results/dataResult.ts
export default class DataResult<T = null> {
    public isSuccess: boolean

    constructor(
        public errors: string[],
        public data: T = <any>null
    ) {
        this.isSuccess = errors.length <= 0
    }

    public static success<T>(value: T): DataResult<T> {
        return new DataResult<T>([], value)
    }

    public static fail(error: string): DataResult {
        return new DataResult([error])
    }

}

export function mapToDataResult<T>(data: any): DataResult<T> {
    return new DataResult(data.errors, data)
}